htslib/hts_endian.h

```diff
@@ -0,0 +1,354 @@
+#if defined(HTS_LITTLE_ENDIAN) && HTS_ALLOW_UNALIGNED != 0
+    *((uint16_u *) buf) = val;
+#else
+    buf[0] = val & 0xff;
+    buf[1] = (val >> 8) & 0xff;
+#endif
+}
+
+/// Store a uint32_t value in little-endian byte order
+/** @param val The value to store
+ *  @param buf Where to store it (may be unaligned)
+ */
+static inline void u32_to_le(uint32_t val, uint8_t *buf) {
+#if defined(HTS_LITTLE_ENDIAN) && HTS_ALLOW_UNALIGNED != 0
+    *((uint32_u *) buf) = val;
+#else
+    buf[0] = val & 0xff;
+    buf[1] = (val >> 8) & 0xff;
+    buf[2] = (val >> 16) & 0xff;
+    buf[3] = (val >> 24) & 0xff;
+#endif
+}
+
+/// Store a uint64_t value in little-endian byte order
+/** @param val The value to store
+ *  @param buf Where to store it (may be unaligned)
+ */
+static inline void u64_to_le(uint64_t val, uint8_t *buf) {
+#if defined(HTS_LITTLE_ENDIAN) && HTS_ALLOW_UNALIGNED != 0
+    *((uint64_u *) buf) = val;
+#else
+    buf[0] = val & 0xff;
+    buf[1] = (val >> 8) & 0xff;
+    buf[2] = (val >> 16) & 0xff;
+    buf[3] = (val >> 24) & 0xff;
+    buf[4] = (val >> 32) & 0xff;
+    buf[5] = (val >> 40) & 0xff;
+    buf[6] = (val >> 48) & 0xff;
+    buf[7] = (val >> 56) & 0xff;
+#endif
+}
+
+/* Signed values.  Grab the data as unsigned, then convert to signed without
+ * triggering undefined behaviour.  On any sensible platform, the conversion
+ * should optimise away to nothing.
+ */
+
+/// Get an int8_t value from an unsigned byte array
+/** @param buf Pointer to source byte array, may be unaligned
+ *  @return A 8 bit signed integer
+ *  The input data is interpreted as 2's complement representation.
+ */
+static inline int8_t le_to_i8(const uint8_t *buf) {
+    return *buf < 0x80 ? *buf : -((int8_t) (0xff - *buf)) - 1;
+}
+
+/// Get an int16_t value from an unsigned byte array
+/** @param buf Pointer to source byte array, may be unaligned
+ *  @return A 16 bit signed integer
+ *  The input data is interpreted as 2's complement representation in
+ *  little-endian byte order.
+ */
+static inline int16_t le_to_i16(const uint8_t *buf) {
+    uint16_t v = le_to_u16(buf);
+    return v < 0x8000 ? v : -((int16_t) (0xffff - v)) - 1;
+}
+
+/// Get an int32_t value from an unsigned byte array
+/** @param buf Pointer to source byte array, may be unaligned
+ *  @return A 32 bit signed integer
+ *  The input data is interpreted as 2's complement representation in
+ *  little-endian byte order.
+ */
+static inline int32_t le_to_i32(const uint8_t *buf) {
+    uint32_t v = le_to_u32(buf);
+    return v < 0x80000000U ? v : -((int32_t) (0xffffffffU - v)) - 1;
+}
+
+/// Get an int64_t value from an unsigned byte array
+/** @param buf Pointer to source byte array, may be unaligned
+ *  @return A 64 bit signed integer
+ *  The input data is interpreted as 2's complement representation in
+ *  little-endian byte order.
+ */
+static inline int64_t le_to_i64(const uint8_t *buf) {
+    uint64_t v = le_to_u64(buf);
+    return (v < 0x8000000000000000ULL
```

Comment:
Actually maybe not an issue.  ULL is long-long which is defined to be at *least* 64-bits, so maybe larger and inefficient, but always big enough to use in conjunction with a uint64_t expression without overflow.